Fix compatibility with Java Plugin 4.0

The Java plugin 4.0 slightly changed its API in order to align with new SonarQube LTS API. The Ccobertura plugin, relying on the Java Analyzer, has to adapt consequently, as analyzes will fail. 

For the change of API, see line [Line #40](https://github.com/SonarSource/sonar-java/blob/5bb12a036d6001ed0e5853a39be21929f29a31ad/java-frontend/src/main/java/org/sonar/plugins/java/api/JavaResourceLocator.java#L40) of the Java plugin. Prior to the change, the method `findResourceByClassName()` was returning a `org.sonar.api.resources.Resource`. Now it is returning an `org.sonar.api.batch.fs.InputFile`.

The method is used in the [CoberturaReportParser, Line #82](https://github.com/SonarQubeCommunity/sonar-cobertura/blob/c081ed4ca610e24d4ec95edc3bbd61c9dab93ba7/src/main/java/org/sonar/plugins/cobertura/CoberturaReportParser.java#L82).

```
[ERROR] Failed to execute goal org.sonarsource.scanner.maven:sonar-maven-plugin:3.1-build231:sonar (default-cli) 
on project HeadShot-java8: Execution default-cli of goal org.sonarsource.scanner.maven:sonar-maven-plugin:3.1-build231:sonar failed: 
An API incompatibility was encountered while executing org.sonarsource.scanner.maven:sonar-maven-plugin:3.1-build231:sonar: java.lang.NoSuchMethodError:
org.sonar.plugins.java.api.JavaResourceLocator.findResourceByClassName(Ljava/lang/String;)Lorg/sonar/api/resources/Resource;

...

Caused by: java.lang.NoSuchMethodError: org.sonar.plugins.java.api.JavaResourceLocator.findResourceByClassName(Ljava/lang/String;)Lorg/sonar/api/resources/Resource;
```

--- a/.travis.yml
+++ b/.travis.yml
@@ -2,14 +2,14 @@ language: java
 sudo: false
 
 jdk:
-  - oraclejdk7
+  - oraclejdk8
 
 install: true
 script: ./travis.sh
 env:
   matrix:
   - TEST=ci
-  - TEST=plugin SQ_VERSION=5.5
+  - TEST=plugin SQ_VERSION=5.6
 
 matrix:
   fast_finish: true

--- a/pom.xml
+++ b/pom.xml
@@ -79,6 +79,16 @@
       <artifactId>guava</artifactId>
       <version>19.0</version>
     </dependency>
+    <dependency>
+      <groupId>commons-io</groupId>
+      <artifactId>commons-io</artifactId>
+      <version>2.4</version>
+    </dependency>
+    <dependency>
+      <groupId>commons-lang</groupId>
+      <artifactId>commons-lang</artifactId>
+      <version>2.6</version>
+    </dependency>
 
     <dependency>
       <groupId>org.sonarsource.sonarqube</groupId>

--- a/pom.xml
+++ b/pom.xml
@@ -3,9 +3,9 @@
   <modelVersion>4.0.0</modelVersion>
 
   <parent>
-    <groupId>org.codehaus.sonar-plugins</groupId>
+    <groupId>org.sonarsource.parent</groupId>
     <artifactId>parent</artifactId>
-    <version>18</version>
+    <version>34</version>
     <relativePath />
   </parent>
 
@@ -41,23 +41,25 @@
   </issueManagement>
 
   <properties>
-    <sonar.version>4.2</sonar.version>
-    <sonar-java.version>2.1</sonar-java.version>
+    <license.title>SonarQube Cobertura Plugin</license.title>
+
+    <sonar.version>5.6</sonar.version>
+    <sonar-java.version>4.0</sonar-java.version>
 
     <maven.test.redirectTestOutputToFile>true</maven.test.redirectTestOutputToFile>
   </properties>
 
   <dependencies>
     <dependency>
-      <groupId>org.codehaus.sonar</groupId>
+      <groupId>org.sonarsource.sonarqube</groupId>
       <artifactId>sonar-plugin-api</artifactId>
       <version>${sonar.version}</version>
       <scope>provided</scope>
     </dependency>
     <dependency>
-      <groupId>org.codehaus.sonar-plugins.java</groupId>
+      <groupId>org.sonarsource.java</groupId>
       <artifactId>sonar-java-plugin</artifactId>
-      <version>2.1</version>
+      <version>${sonar-java.version}</version>
       <type>sonar-plugin</type>
       <scope>provided</scope>
     </dependency>
@@ -73,9 +75,22 @@
     </dependency>
 
     <dependency>
-      <groupId>org.codehaus.sonar</groupId>
+      <groupId>com.google.guava</groupId>
+      <artifactId>guava</artifactId>
+      <version>19.0</version>
+    </dependency>
+
+    <dependency>
+      <groupId>org.sonarsource.sonarqube</groupId>
       <artifactId>sonar-testing-harness</artifactId>
       <version>${sonar.version}</version>
+      <scope>test</scope>
+    </dependency>
+    <dependency>
+      <groupId>org.sonarsource.sonarqube</groupId>
+      <artifactId>sonar-plugin-api</artifactId>
+      <version>${sonar.version}</version>
+      <type>test-jar</type>
       <scope>test</scope>
     </dependency>
     <dependency>
@@ -88,6 +103,12 @@
       <groupId>org.mockito</groupId>
       <artifactId>mockito-all</artifactId>
       <version>1.9.5</version>
+      <scope>test</scope>
+    </dependency>
+    <dependency>
+      <groupId>ch.qos.logback</groupId>
+      <artifactId>logback-classic</artifactId>
+      <version>0.9.30</version>
       <scope>test</scope>
     </dependency>
   </dependencies>
@@ -104,7 +125,7 @@
     </testResources>
     <plugins>
       <plugin>
-        <groupId>org.codehaus.sonar</groupId>
+        <groupId>org.sonarsource.sonar-packaging-maven-plugin</groupId>
         <artifactId>sonar-packaging-maven-plugin</artifactId>
         <configuration>
           <pluginKey>cobertura</pluginKey>

--- a/src/main/java/org/sonar/plugins/cobertura/CoberturaPlugin.java
+++ b/src/main/java/org/sonar/plugins/cobertura/CoberturaPlugin.java
@@ -1,7 +1,7 @@
 /*
  * SonarQube Cobertura Plugin
- * Copyright (C) 2013 SonarSource
- * dev@sonar.codehaus.org
+ * Copyright (C) 2013-2016 SonarSource SA
+ * mailto:contact AT sonarsource DOT com
  *
  * This program is free software; you can redistribute it and/or
  * modify it under the terms of the GNU Lesser General Public
@@ -13,9 +13,9 @@
  * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
  * Lesser General Public License for more details.
  *
- * You should have received a copy of the GNU Lesser General Public
- * License along with this program; if not, write to the Free Software
- * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
  */
 package org.sonar.plugins.cobertura;
 

--- a/src/main/java/org/sonar/plugins/cobertura/CoberturaReportParser.java
+++ b/src/main/java/org/sonar/plugins/cobertura/CoberturaReportParser.java
@@ -91,7 +91,7 @@ public class CoberturaReportParser {
   }
 
   private boolean resourceExists(InputFile file) {
-    return context.getResource(file) != null;
+    return file != null && context.getResource(file) != null;
   }
 
   private static void collectFileMeasures(SMInputCursor clazz, Map<String, CoverageMeasuresBuilder> builderByFilename) throws XMLStreamException {

--- a/src/main/java/org/sonar/plugins/cobertura/CoberturaReportParser.java
+++ b/src/main/java/org/sonar/plugins/cobertura/CoberturaReportParser.java
@@ -1,7 +1,7 @@
 /*
  * SonarQube Cobertura Plugin
- * Copyright (C) 2013 SonarSource
- * dev@sonar.codehaus.org
+ * Copyright (C) 2013-2016 SonarSource SA
+ * mailto:contact AT sonarsource DOT com
  *
  * This program is free software; you can redistribute it and/or
  * modify it under the terms of the GNU Lesser General Public
@@ -13,9 +13,9 @@
  * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
  * Lesser General Public License for more details.
  *
- * You should have received a copy of the GNU Lesser General Public
- * License along with this program; if not, write to the Free Software
- * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
  */
 package org.sonar.plugins.cobertura;
 
@@ -25,6 +25,7 @@ import org.apache.commons.lang.StringUtils;
 import org.codehaus.staxmate.in.SMHierarchicCursor;
 import org.codehaus.staxmate.in.SMInputCursor;
 import org.sonar.api.batch.SensorContext;
+import org.sonar.api.batch.fs.InputFile;
 import org.sonar.api.measures.CoverageMeasuresBuilder;
 import org.sonar.api.measures.Measure;
 import org.sonar.api.resources.Resource;
@@ -79,7 +80,7 @@ public class CoberturaReportParser {
       collectFileMeasures(pack.descendantElementCursor("class"), builderByFilename);
       for (Map.Entry<String, CoverageMeasuresBuilder> entry : builderByFilename.entrySet()) {
         String className = sanitizeFilename(entry.getKey());
-        Resource resource = javaResourceLocator.findResourceByClassName(className);
+        InputFile resource = javaResourceLocator.findResourceByClassName(className);
         if (resourceExists(resource)) {
           for (Measure measure : entry.getValue().createMeasures()) {
             context.saveMeasure(resource, measure);
@@ -89,7 +90,7 @@ public class CoberturaReportParser {
     }
   }
 
-  private boolean resourceExists(Resource file) {
+  private boolean resourceExists(InputFile file) {
     return context.getResource(file) != null;
   }
 

--- a/src/main/java/org/sonar/plugins/cobertura/CoberturaSensor.java
+++ b/src/main/java/org/sonar/plugins/cobertura/CoberturaSensor.java
@@ -1,7 +1,7 @@
 /*
  * SonarQube Cobertura Plugin
- * Copyright (C) 2013 SonarSource
- * dev@sonar.codehaus.org
+ * Copyright (C) 2013-2016 SonarSource SA
+ * mailto:contact AT sonarsource DOT com
  *
  * This program is free software; you can redistribute it and/or
  * modify it under the terms of the GNU Lesser General Public
@@ -13,9 +13,9 @@
  * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
  * Lesser General Public License for more details.
  *
- * You should have received a copy of the GNU Lesser General Public
- * License along with this program; if not, write to the Free Software
- * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
  */
 package org.sonar.plugins.cobertura;
 

--- a/src/main/java/org/sonar/plugins/cobertura/package-info.java
+++ b/src/main/java/org/sonar/plugins/cobertura/package-info.java
@@ -1,7 +1,7 @@
 /*
  * SonarQube Cobertura Plugin
- * Copyright (C) 2013 SonarSource
- * dev@sonar.codehaus.org
+ * Copyright (C) 2013-2016 SonarSource SA
+ * mailto:contact AT sonarsource DOT com
  *
  * This program is free software; you can redistribute it and/or
  * modify it under the terms of the GNU Lesser General Public
@@ -13,9 +13,8 @@
  * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
  * Lesser General Public License for more details.
  *
- * You should have received a copy of the GNU Lesser General Public
- * License along with this program; if not, write to the Free Software
- * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
  */
-
 @javax.annotation.ParametersAreNonnullByDefault package org.sonar.plugins.cobertura;

--- a/src/test/java/org/sonar/plugins/cobertura/CoberturaPluginTest.java
+++ b/src/test/java/org/sonar/plugins/cobertura/CoberturaPluginTest.java
@@ -1,7 +1,7 @@
 /*
  * SonarQube Cobertura Plugin
- * Copyright (C) 2013 SonarSource
- * dev@sonar.codehaus.org
+ * Copyright (C) 2013-2016 SonarSource SA
+ * mailto:contact AT sonarsource DOT com
  *
  * This program is free software; you can redistribute it and/or
  * modify it under the terms of the GNU Lesser General Public
@@ -13,9 +13,9 @@
  * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
  * Lesser General Public License for more details.
  *
- * You should have received a copy of the GNU Lesser General Public
- * License along with this program; if not, write to the Free Software
- * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
  */
 package org.sonar.plugins.cobertura;
 

--- a/src/test/java/org/sonar/plugins/cobertura/CoberturaSensorTest.java
+++ b/src/test/java/org/sonar/plugins/cobertura/CoberturaSensorTest.java
@@ -1,7 +1,7 @@
 /*
  * SonarQube Cobertura Plugin
- * Copyright (C) 2013 SonarSource
- * dev@sonar.codehaus.org
+ * Copyright (C) 2013-2016 SonarSource SA
+ * mailto:contact AT sonarsource DOT com
  *
  * This program is free software; you can redistribute it and/or
  * modify it under the terms of the GNU Lesser General Public
@@ -13,9 +13,9 @@
  * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
  * Lesser General Public License for more details.
  *
- * You should have received a copy of the GNU Lesser General Public
- * License along with this program; if not, write to the Free Software
- * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
  */
 package org.sonar.plugins.cobertura;
 
@@ -24,6 +24,7 @@ import org.fest.assertions.Assertions;
 import org.junit.Before;
 import org.junit.Test;
 import org.sonar.api.batch.SensorContext;
+import org.sonar.api.batch.fs.InputFile;
 import org.sonar.api.config.Settings;
 import org.sonar.api.measures.CoreMetrics;
 import org.sonar.api.measures.Measure;
@@ -58,6 +59,7 @@ public class CoberturaSensorTest {
   private PathResolver pathResolver;
   private Settings settings;
   private JavaResourceLocator javaResourceLocator;
+  private InputFile inputFile;
   private Resource resource;
   private ModuleFileSystem fs;
 
@@ -69,6 +71,7 @@ public class CoberturaSensorTest {
     settings = new Settings();
     javaResourceLocator = mock(JavaResourceLocator.class);
     sensor = new CoberturaSensor(fs, pathResolver, settings, javaResourceLocator);
+    inputFile = mock(InputFile.class);
     resource = mock(Resource.class);
   }
 
@@ -136,22 +139,22 @@ public class CoberturaSensorTest {
 
   @Test
   public void collectFileLineCoverage() throws URISyntaxException {
-    when(javaResourceLocator.findResourceByClassName("org.apache.commons.chain.config.ConfigParser")).thenReturn(resource);
-    when(context.getResource(any(Resource.class))).thenReturn(resource);
+    when(javaResourceLocator.findResourceByClassName("org.apache.commons.chain.config.ConfigParser")).thenReturn(inputFile);
+    when(context.getResource(inputFile)).thenReturn(resource);
     sensor.parseReport(getCoverageReport(), context);
 
-    verify(context).saveMeasure(eq(resource), argThat(new IsMeasure(CoreMetrics.LINES_TO_COVER, 30.0)));
-    verify(context).saveMeasure(eq(resource), argThat(new IsMeasure(CoreMetrics.UNCOVERED_LINES, 5.0)));
+    verify(context).saveMeasure(eq(inputFile), argThat(new IsMeasure(CoreMetrics.LINES_TO_COVER, 30.0)));
+    verify(context).saveMeasure(eq(inputFile), argThat(new IsMeasure(CoreMetrics.UNCOVERED_LINES, 5.0)));
   }
 
   @Test
   public void collectFileBranchCoverage() throws URISyntaxException {
-    when(javaResourceLocator.findResourceByClassName("org.apache.commons.chain.config.ConfigParser")).thenReturn(resource);
-    when(context.getResource(any(Resource.class))).thenReturn(resource);
+    when(javaResourceLocator.findResourceByClassName("org.apache.commons.chain.config.ConfigParser")).thenReturn(inputFile);
+    when(context.getResource(inputFile)).thenReturn(resource);
     sensor.parseReport(getCoverageReport(), context);
 
-    verify(context).saveMeasure(eq(resource), argThat(new IsMeasure(CoreMetrics.CONDITIONS_TO_COVER, 6.0)));
-    verify(context).saveMeasure(eq(resource), argThat(new IsMeasure(CoreMetrics.UNCOVERED_CONDITIONS, 2.0)));
+    verify(context).saveMeasure(eq(inputFile), argThat(new IsMeasure(CoreMetrics.CONDITIONS_TO_COVER, 6.0)));
+    verify(context).saveMeasure(eq(inputFile), argThat(new IsMeasure(CoreMetrics.UNCOVERED_CONDITIONS, 2.0)));
   }
 
   @Test
@@ -185,21 +188,21 @@ public class CoberturaSensorTest {
     File coverage = new File(getClass().getResource(
         "/org/sonar/plugins/cobertura/CoberturaSensorTest/shouldInsertCoverageAtFileLevel/coverage.xml").toURI());
     when(resource.getName()).thenReturn("org.sonar.samples.InnerClass");
-    when(javaResourceLocator.findResourceByClassName("org.sonar.samples.InnerClass")).thenReturn(resource);
-    when(javaResourceLocator.findResourceByClassName("org.sonar.samples.InnerClass$InnerClassInside")).thenReturn(resource);
-    when(javaResourceLocator.findResourceByClassName("org.sonar.samples.PrivateClass")).thenReturn(resource);
+    when(javaResourceLocator.findResourceByClassName("org.sonar.samples.InnerClass")).thenReturn(inputFile);
+    when(javaResourceLocator.findResourceByClassName("org.sonar.samples.InnerClass$InnerClassInside")).thenReturn(inputFile);
+    when(javaResourceLocator.findResourceByClassName("org.sonar.samples.PrivateClass")).thenReturn(inputFile);
 
-    when(context.getResource(any(Resource.class))).thenReturn(resource);
+    when(context.getResource(inputFile)).thenReturn(resource);
     sensor.parseReport(coverage, context);
 
-    verify(context).saveMeasure(eq(resource),
+    verify(context).saveMeasure(eq(inputFile),
         argThat(new IsMeasure(CoreMetrics.LINES_TO_COVER, 35.0)));
-    verify(context).saveMeasure(eq(resource),
+    verify(context).saveMeasure(eq(inputFile),
         argThat(new IsMeasure(CoreMetrics.UNCOVERED_LINES, 22.0)));
 
-    verify(context).saveMeasure(eq(resource),
+    verify(context).saveMeasure(eq(inputFile),
         argThat(new IsMeasure(CoreMetrics.CONDITIONS_TO_COVER, 4.0)));
-    verify(context).saveMeasure(eq(resource),
+    verify(context).saveMeasure(eq(inputFile),
         argThat(new IsMeasure(CoreMetrics.UNCOVERED_CONDITIONS, 3.0)));
 
     verify(context, never()).saveMeasure(
@@ -230,7 +233,7 @@ public class CoberturaSensorTest {
 
     verify(context)
         .saveMeasure(
-            eq(resource),
+            eq(inputFile),
             argThat(new IsMeasure(
                 CoreMetrics.COVERAGE_LINE_HITS_DATA,
                 "22=2;25=0;26=0;29=0;30=0;31=0;34=1;35=1;36=1;37=0;39=1;41=1;44=2;46=1;47=1;50=0;51=0;52=0;53=0;55=0;57=0;60=0;61=0;64=1;71=1;73=1;76=0;77=0;80=0;81=0;85=0;87=0;91=0;93=0;96=1")));
@@ -238,12 +241,12 @@ public class CoberturaSensorTest {
 
   @Test
   public void collectFileLineHitsData() throws URISyntaxException {
-    when(javaResourceLocator.findResourceByClassName("org.apache.commons.chain.impl.CatalogBase")).thenReturn(resource);
-    when(context.getResource(any(Resource.class))).thenReturn(resource);
+    when(javaResourceLocator.findResourceByClassName("org.apache.commons.chain.impl.CatalogBase")).thenReturn(inputFile);
+    when(context.getResource(inputFile)).thenReturn(resource);
 
     sensor.parseReport(getCoverageReport(), context);
     verify(context).saveMeasure(
-        eq(resource),
+        eq(inputFile),
         argThat(new IsMeasure(CoreMetrics.COVERAGE_LINE_HITS_DATA,
             "48=117;56=234;66=0;67=0;68=0;84=999;86=999;98=318;111=18;121=0;122=0;125=0;126=0;127=0;128=0;131=0;133=0")));
   }
@@ -251,11 +254,11 @@ public class CoberturaSensorTest {
   @Test
   public void shouldNotCountTwiceAnonymousClasses() throws URISyntaxException {
     File coverage = new File(getClass().getResource("/org/sonar/plugins/cobertura/CoberturaSensorTest/shouldNotCountTwiceAnonymousClasses.xml").toURI());
-    when(javaResourceLocator.findResourceByClassName("org.sonar.samples.MyFile")).thenReturn(resource);
-    when(context.getResource(any(Resource.class))).thenReturn(resource);
+    when(javaResourceLocator.findResourceByClassName("org.sonar.samples.MyFile")).thenReturn(inputFile);
+    when(context.getResource(inputFile)).thenReturn(resource);
     sensor.parseReport(coverage, context);
 
-    verify(context).saveMeasure(eq(resource), //argThat(new IsResource(Scopes.FILE, Qualifiers.CLASS, "org.sonar.samples.MyFile")),
+    verify(context).saveMeasure(eq(inputFile), //argThat(new IsResource(Scopes.FILE, Qualifiers.CLASS, "org.sonar.samples.MyFile")),
         argThat(new IsMeasure(CoreMetrics.LINES_TO_COVER, 5.0))); // do not count line 26 twice
   }
 

--- a/travis.sh
+++ b/travis.sh
@@ -20,7 +20,7 @@ plugin)
   mvn package -T2 -Dsource.skip=true -Denforcer.skip=true -Danimal.sniffer.skip=true -Dmaven.test.skip=true
 
   cd its/$TEST
-  mvn package -Dsonar.runtimeVersion="$SQ_VERSION" -DjavaVersion="LATEST_RELEASE" -Dmaven.test.redirectTestOutputToFile=false
+  mvn package -B -Dsonar.runtimeVersion="$SQ_VERSION" -DjavaVersion="LATEST_RELEASE" -Dmaven.test.redirectTestOutputToFile=false
   ;;
 
 *)

